bench/tpcc: use 3-node cluster, refactor, and create cleaner profiles

mgartner opened the pull request:
#### bench/tpcc: use 3-node cluster

BenchmarkTPCC now uses a 3-node cluster instead of a single-node
cluster. Starting-up a 3-node cluster takes several seconds, so the
benchmark now only does this once for all implementations (literal or
optimized) and mixes, rather than once per implemenation/mix
combination.

Also, we no longer generate a reusable store directory. Instead, we use
in-memory stores and initialize the TPC-C workload data each time the
benchmark is run. This provides a simpler user experience and the benchmark
runs just as quickly as reusing pre-initialized stores.

Release note: None

#### bench/tpcc: set cluster settings

Release note: None

#### bench/tpcc: move client code to same process

Moving the workload client to the same process vastly simplifies the
benchmark.

Release note: None

#### bench/tpcc: collect CPU and memory profile that exclude setup

CPU and heap profiles can be collected while running the benchmarks by
passing the "-profile" flag with a prefix for the profile file names.
The profile names will include the benchmark names, e.g.,
`foo_tpcc_literal_new_order.cpu` and `foo_tpcc_literal_new_order.mem`
will be created when running the benchmark:

    ./dev bench pkg/bench/tpcc -f BenchmarkTPCC/literal/new_order \
      --test-args '-test.benchtime=30s -profile=foo'

These profiles will omit CPU samples and allocations made during the setup
phase of the benchmark, unlike profiles collected with the `-test.cpuprofile`
and `-test.memprofile` flags.

Release note: None

#### tpcc: disable stdout in benchmark

The workload plumbing is cli-centric, but here we're using tpcc as a library.
While invoking it in this way doesn't trigger any workload-level logging
(behind the `--display-format` flag), tpcc has its own logging during data
loading. This commit makes it possible to override this logging, and does so in
BenchmarkTPCC.

In an ideal world, there would be configuration options for this sort of thing,
but this would be a large and painful refactor.

#### bench/tpcc: refactor and disable local client optimization

Release note: None

#### bench/tpcc: round-robin queries to all three nodes

Release note: None

#### bench/tpcc: use import data loader

Using the import data loader instead of the insert data loader speeds up
the setup time for the benchmark significantly.

Release note: None

#### bench/tpcc: run each mix on its own cluster

Each run of a benchmark is now run in its own cluster to make each
run independent of each other and avoid skewing results.

Release note: None

#### bench/tpcc: remove --profile flag

The `--profile` flag is no longer supported. Instead, CPU and heap
profiles will be collected if `-test.cpuprofile` or `-test.memprofile`
are provided. The benchmark will "hijack" the standard profiles and
create new profiles that omit CPU samples and allocations made during
the setup phase of the benchmark.

The profile names will include the prefix of the profile flags and the
benchmark names. For example, the profiles "foo_tpcc_literal_new_order.cpu"
and "foo_tpcc_literal_new_order.mem" will be created when running the
benchmark:

    ./dev bench pkg/bench/tpcc -f BenchmarkTPCC/literal/new_order \
      --test-args '-test.cpuprofile=foo.cpu -test.memprofile=foo.mem'

Note that the "foo.cpu" file will not be created because we must stop
the global CPU profiler in order to collect profiles that omit setup
samples. The "foo.mem" file will created and include all allocations
made during the entire duration of the benchmark.

Release note: None

#### bench/benchprof: add benchprof package

The benchprof package has been created to provide utilities for creating
CPU and memory profiles in benchmarks.

Release note: None

#### bench/benchprof: support mutex profiles

Release note: None

mgartner committed:
bench/tpcc: use 3-node cluster

BenchmarkTPCC now uses a 3-node cluster instead of a single-node
cluster. Starting-up a 3-node cluster takes several seconds, so the
benchmark now only does this once for all implementations (literal or
optimized) and mixes, rather than once per implemenation/mix
combination.

Also, we no longer generate a reusable store directory. Instead, we use
in-memory stores and initialize the TPC-C workload data each time the
benchmark is run. This provides a simpler user experience and the benchmark
runs just as quickly as reusing pre-initialized stores.

Release note: None

mgartner committed:
bench/tpcc: set cluster settings

Release note: None

mgartner committed:
bench/tpcc: move client code to same process

Moving the workload client to the same process vastly simplifies the
benchmark.

Release note: None

mgartner committed:
bench/tpcc: collect CPU and memory profile that exclude setup

CPU and heap profiles can be collected while running the benchmarks by
passing the "-profile" flag with a prefix for the profile file names.
The profile names will include the benchmark names, e.g.,
`foo_tpcc_literal_new_order.cpu` and `foo_tpcc_literal_new_order.mem`
will be created when running the benchmark:

    ./dev bench pkg/bench/tpcc -f BenchmarkTPCC/literal/new_order \
      --test-args '-test.benchtime=30s -profile=foo'

These profiles will omit CPU samples and allocations made during the setup
phase of the benchmark, unlike profiles collected with the `-test.cpuprofile`
and `-test.memprofile` flags.

Release note: None

mgartner commented:
I'm pretty happy with this iteration, and I think it works well. I've completely removed the reusable store directories—they were causing lots of issues without a huge benefit. I think the UX is pretty nice now, but if you want to kick the tires I'd appreciate it.

tbg committed:
tpcc: disable stdout in benchmark

The workload plumbing is cli-centric, but here we're using tpcc as a library.
While invoking it in this way doesn't trigger any workload-level logging
(behind the `--display-format` flag), tpcc has its own logging during data
loading. This commit makes it possible to override this logging, and does so in
BenchmarkTPCC.

In an ideal world, there would be configuration options for this sort of thing,
but this would be a large and painful refactor.

tbg reviewed (approved):
Flushing comments. Now taking it out on the road 🚗

tbg commented:
I wanted to understand the (possibly negative) impact of reusing the server, so I ran with `-test.count=100`:

```
./dev bench pkg/bench/tpcc --filter BenchmarkTPCC/optimized/new_order --test-args '-test.cpu=12 -test.count=100 -test.benchtime=200x' 2>&1 | tee bench.txt
```

The first time around, on the 14th iteration it failed with a memory budget exceeded error:

<details><summary>Details</summary>
<p>


```
BenchmarkTPCC/optimized/new_order-12            --- FAIL: BenchmarkTPCC/optimized/new_order-12
    bench_test.go:115: error printed in newOrder: update stock failed: ERROR: root: memory budget exceeded: 10240 bytes requested, 268435456 currently allocated, 268435456 bytes in budget (SQ
BenchmarkTPCC/optimized/new_order-12            --- FAIL: BenchmarkTPCC/optimized/new_order
    bench_test.go:115: error printed in newOrder: failed to connect to `host=127.0.0.1 user=root database=tpcc`: server error (ERROR: root: memory budget exceeded: 102400 bytes requested, 268
(SQLSTATE 53200))
--- FAIL: BenchmarkTPCC/optimized/new_order-12
    bench_test.go:115: error printed in newOrder: failed to connect to `host=127.0.0.1 user=root database=tpcc`: server error (ERROR: root: memory budget exceeded: 102400 bytes requested, 268
(SQLSTATE 53200))
    bench_test.go:115: error printed in newOrder: failed to connect to `host=127.0.0.1 user=root database=tpcc`: server error (ERROR: root: memory budget exceeded: 102400 bytes requested, 268
(SQLSTATE 53200))
BenchmarkTPCC/optimized/new_order-12            panic: assignment to entry in nil map

goroutine 158562 [running]:
github.com/cockroachdb/cockroach/pkg/security/clientcert.(*Cache).Upsert(0xc00134d8c0, {0x68f9598, 0xc0111ad040}, {0xc0065688f0, 0x4}, {0xc0475c3680, 0x27}, 0x71732dbe)
        pkg/security/clientcert/cert_expiry_cache.go:116 +0x374
github.com/cockroachdb/cockroach/pkg/security.(*CertificateManager).MaybeUpsertClientExpiration(...)
        pkg/security/certificate_manager.go:222
github.com/cockroachdb/cockroach/pkg/security.UserAuthCertHook.func1({0x68f9598, 0xc0111ad040}, {0xc0065688f0, 0x4}, 0x0?)
        pkg/security/auth.go:356 +0x4b4
github.com/cockroachdb/cockroach/pkg/sql/pgwire.authCert.func1({0x68f9598, 0xc0111ad040}, {0xc0065688f0, 0x4}, 0x1, 0xc01de9f0f1?, 0x0)
        pkg/sql/pgwire/auth_methods.go:460 +0x212
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*AuthBehaviors).Authenticate(0xc03807e700?, {0x68f9598?, 0xc0111ad040?}, {0xc0065688f0?, 0xc0111ad040?}, 0x8?, 0xc0065688f0?, 0x4?)
        pkg/sql/pgwire/auth_behaviors.go:54 +0x43
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).handleAuthentication(0xc047e0e408, {0x68f9598, 0xc0111ad040}, {0x695a148, 0xc0059e9e40}, {0x0, 0x4, {0x1, {0x5c323ef, 0x3}, ...}, ...},
        pkg/sql/pgwire/auth.go:185 +0x745
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands(0xc047e0e408, {0x68f9598, 0xc0111ad040}, {0x0, 0x4, {0x1, {0x5c323ef, 0x3}, {0xc006568980, 0xf}, ...}, ...}, ...)
        pkg/sql/pgwire/conn.go:205 +0x185
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func4()
        pkg/sql/pgwire/server.go:1209 +0xec
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl in goroutine 158521
        pkg/sql/pgwire/server.go:1206 +0x5f5
I250606 08:37:49.554500 1 (gostd) testmain.go:87  [-] 1  Test //pkg/bench/tpcc:tpcc_test exited with error code 2
```

</p>
</details> 

Assuming that we were leaking memory budget somewhere, I re-ran, expecting a failure at the same iteration. But this time it  kept going for significantly longer, failing only on the 57th iteration.

mgartner commented on pkg/bench/tpcc/bench_test.go:
```diff
@@ -25,14 +33,33 @@ import (
 )
 
-// BenchmarkTPCC runs TPC-C transactions against a single warehouse.
+var profileFlag = flag.String("profile", "",
```
> Is there a reason you're doing this differently from BenchmarkSysbench, which piggybacks on the existing flags for profiles?

Mostly for the benefits for the CPU profiles that you mentioned. I also like that it's much less verbose.

mgartner replied:
Also, I don't see any reason why someone would want to profile while running `benchdiff`—it would likely affect results too—so I wasn't concerned about the fact that it wouldn't work in that situation.

mgartner commented on pkg/bench/tpcc/bench_test.go:
```diff
@@ -31,14 +32,11 @@ const (
-// not included in CPU and heap profiles.
+// BenchmarkTPCC runs TPC-C transactions against a single warehouse.
 func BenchmarkTPCC(b *testing.B) {
-	defer leaktest.AfterTest(b)()
```
I think only for tests. Other benchmarks, like sysbench, don't use this. And there's some goroutine leaks with a test cluster that were causing the benchmark to fail. If you add this line to the sysbench micro-benchmark you get failures as well. I couldn't quickly solve it so I kicked that can down the road for now.

mgartner committed:
bench/tpcc: refactor and disable local client optimization

Release note: None

mgartner commented on pkg/bench/tpcc/bench_test.go:
```diff
@@ -89,43 +85,69 @@ func run(b *testing.B, storeDir string, workloadFlags []string) {
+	}
+
+	var l workloadsql.InsertsDataLoader
+	if _, err := workloadsql.Setup(ctx, c.ServerConn(0), gen, l); err != nil {
```
Good idea. I'll give that a try.

mgartner commented:
Still working on some more changes, so no need to re-review yet. Responded to some inline comments above and below.

---

>Based on this argument, I would argue that cluster reuse is not a good idea, since we can achieve that behavior already by passing -test.count=1. We should strive to make independent benchmark runs as disjoint from each other as possible.

IMO not reusing the cluster makes the benchmark so slow to run that it's essentially useless. Running a benchmark with a single `new_order` mix with no extended `-test.benchtime` and `--count 5` on my mac:

```
./dev bench pkg/bench/tpcc -f BenchmarkTPCC/literal/order_status --count 5 --timeout 3000s
```

Method | Time
------ | ----
Reusing cluster | 32.3 seconds
Not Reusing cluster | 500.5 seconds

I think a lot of difference is due to the silly way that Go's benchmark tool runs a benchmark first for a single iteration (`b.N=1`) to determine a good `b.N` and then re-runs its. So by moving the cluster initialization into the lowest of the `b.Run`s, you pay for the initialize at least twice per `--count`, assuming you don't set `-test.benchtime` manually. But that's how most folks will run it to get a gut-check on their changes. I don't think it's kind to impose that type of overhead by default. 30 seconds vs 500 seconds is the difference between you getting up to refill your coffee and getting back to work immediately vs you getting distracted by Slack/email/etc. and coming back to your optimizations several minutes later with context paged out.

> The first time around, on the 14th iteration it failed with a memory budget exceeded error:

This might be related to the goroutine leak, but not sure without further investigation.

mgartner committed:
bench/tpcc: round-robin queries to all three nodes

Release note: None

tbg commented:
> IMO not reusing the cluster makes the benchmark so slow to run that it's essentially useless. Running a benchmark with a single new_order mix with no extended -test.benchtime and --count 5 on my mac:

Why is data setup so slow? `BenchmarkSysbench` sets up ten tables at 10k rows each, and this takes no more than a few seconds. Is there something to be fixed here? Sadly a TPCC 1WH dataset has just shy of 600k rows, so it isn't "small". For a the purpose of this benchmark, it would make sense to scale it down, but unfortunately - that would require nontrivial changes to the workload. Is the import using the CSV data loader? Maybe that will speed it up?

If the benchmark is either too slow or runs aren't independent, aren't both options problematic?

mgartner committed:
bench/tpcc: use import data loader

Using the import data loader instead of the insert data loader speeds up
the setup time for the benchmark significantly.

Release note: None

mgartner committed:
bench/tpcc: run each mix on its own cluster

Each run of a benchmark is now run in its own cluster to make each
run independent of each other and avoid skewing results.

Release note: None

mgartner committed:
bench/tpcc: remove --profile flag

The `--profile` flag is no longer supported. Instead, CPU and heap
profiles will be collected if `-test.cpuprofile` or `-test.memprofile`
are provided. The benchmark will "hijack" the standard profiles and
create new profiles that omit CPU samples and allocations made during
the setup phase of the benchmark.

The profile names will include the prefix of the profile flags and the
benchmark names. For example, the profiles "foo_tpcc_literal_new_order.cpu"
and "foo_tpcc_literal_new_order.mem" will be created when running the
benchmark:

    ./dev bench pkg/bench/tpcc -f BenchmarkTPCC/literal/new_order \
      --test-args '-test.cpuprofile=foo.cpu -test.memprofile=foo.mem'

Note that the "foo.cpu" file will not be created because we must stop
the global CPU profiler in order to collect profiles that omit setup
samples. The "foo.mem" file will created and include all allocations
made during the entire duration of the benchmark.

Release note: None

mgartner committed:
bench/benchprof: add benchprof package

The benchprof package has been created to provide utilities for creating
CPU and memory profiles in benchmarks.

Release note: None

mgartner committed:
bench/benchprof: support mutex profiles

Release note: None

mgartner commented:
@tbg please take another look. I've added a `benchprof` package to make it easy to reuse the profiling logic.

tbg commented on pkg/bench/benchprof/benchprof.go:
```diff
@@ -23,8 +23,35 @@ import (
-func (f StopFn) Stop(b testing.TB) {
-	f(b)
+func (f StopFn) Stop(tb testing.TB) {
+	f(tb)
```
This is fine, but you can also `if f != nil { f(tbg) }` and then you can use `nil` as a noop stop function, which you could use in 2-3 places here.

tbg commented on pkg/bench/benchprof/benchprof.go:
```diff
@@ -126,25 +149,75 @@ func StartMemProfile(tb testing.TB) StopFn {
-	if outputDir != "" {
-		memProfFile = filepath.Join(outputDir, memProfFile)
+	if mutexProfFile == "" {
+		// No CPU profile requested, do not collect a mutex profile.
```
Comment is off :)

mgartner replied:
Fixed.

tbg commented on pkg/bench/benchprof/benchprof.go:
```diff
@@ -126,25 +149,75 @@ func StartMemProfile(tb testing.TB) StopFn {
-	var outputDir string
-	if err := sniffarg.DoEnv("test.outputdir", &outputDir); err != nil {
+	// Create a new profile file.
+	fileName := profileFileName(tb, dir, prefix, "mem")
```
"allocs" could be more appropriate but this works too.

tbg reviewed (approved):
Very nice - thank you for going the extra mile!
